### ISSUE TITLE
Failed to create NNI experiment, error:  AttributeError: 'dict' object has no attribute 'name'

### ISSUE THREAD

**Issue by sunze992** [microsoft/nni#5457]
**Describe the issue**:
But when I created the NNI experiment using the official example of Hyperparameter Optimization, I received the following error information:

```
Traceback (most recent call last):
  File "/home/sunze/anaconda3/bin/nnictl", line 8, in <module>
    sys.exit(parse_args())
  File "/home/sunze/anaconda3/lib/python3.9/site-packages/nni/tools/nnictl/nnictl.py", line 497, in parse_args
    args.func(args)
  File "/home/sunze/anaconda3/lib/python3.9/site-packages/nni/tools/nnictl/launcher.py", line 91, in create_experiment
    exp.start(port, debug, RunMode.Detach)
  File "/home/sunze/anaconda3/lib/python3.9/site-packages/nni/experiment/experiment.py", line 135, in start
    self._start_impl(port, debug, run_mode, None, [])
  File "/home/sunze/anaconda3/lib/python3.9/site-packages/nni/experiment/experiment.py", line 94, in _start_impl
    config = self.config.canonical_copy()
  File "/home/sunze/anaconda3/lib/python3.9/site-packages/nni/experiment/config/base.py", line 166, in canonical_copy
    canon._canonicalize([])
  File "/home/sunze/anaconda3/lib/python3.9/site-packages/nni/experiment/config/experiment_config.py", line 124, in _canonicalize
    if algo is not None and algo.name == '_none_':  # type: ignore
AttributeError: 'dict' object has no attribute 'name'
```

**The NNI source code at the wrong place is as follows**:
```
def _canonicalize(self, _parents):
        if self.log_level is None:
            self.log_level = 'debug' if self.debug else 'info'
        self.tuner_gpu_indices = utils.canonical_gpu_indices(self.tuner_gpu_indices)

        for algo_type in ['tuner', 'assessor', 'advisor']:
            algo = getattr(self, algo_type)
            # TODO: need a more universal solution for similar problems
            if isinstance(algo, dict):
                # the base class should have converted it to `_AlgorithmConfig` if feasible
                # it is a `dict` here means an exception was raised during the convertion attempt
                # we do the convertion again to show user the error message
                _AlgorithmConfig(**algo)  # pylint: disable=not-a-mapping

            if algo is not None and algo.name == '_none_':  # type: ignore
                setattr(self, algo_type, None)


```
**I found ‘_algo_’ in code should be type ‘__AlgorithmConfig_’, but when I was running, I got the ‘_dict_’ type. I think this is the root of the problem, but I don't know how to solve it. My other Ubuntu server can run normally with the same environment, but this machine reports the error on the above error. So I want to ask what to do in this situation, I can't think of other solutions anymore. Thank you very much for your answer!**

**Environment**:
- NNI version: 2.9
- Training service (local|remote|pai|aml|etc): local
- Client OS: Ubuntu 20.04
- Python version: 3.9.13
- PyTorch version: 1.12.0
- Is conda/virtualenv/venv used?: conda used
- Is running in Docker?: No

**Configuration**:
```
experimentName: demo

searchSpace:
  features:
    _type: choice
    _value: [ 128, 256, 512, 1024 ]
  lr:
    _type: loguniform
    _value: [ 0.0001, 0.1 ]
  momentum:
    _type: uniform
    _value: [ 0, 1 ]

trialCommand: python model.py
trialCodeDirectory: .

trialGpuNumber: 1
trialConcurrency: 4
maxTrialNumber: 80

tuner:
  name: TPE
  classArgs:
    optimize_mode: maximize

trainingService:
  platform: local
  useActiveGpu: True
  maxTrialNumberPerGpu: 1

```

**The very simple code of model.py**:
```
import nni
import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision.transforms import ToTensor

params = {
    'features': 512,
    'lr': 0.001,
    'momentum': 0,
}

optimized_params = nni.get_next_parameter()

params.update(optimized_params)


training_data = datasets.FashionMNIST(root="data", train=True, download=True, transform=ToTensor())
test_data = datasets.FashionMNIST(root="data", train=False, download=True, transform=ToTensor())

batch_size = 64

train_dataloader = DataLoader(training_data, batch_size=batch_size)
test_dataloader = DataLoader(test_data, batch_size=batch_size)

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using {device} device")

class NeuralNetwork(nn.Module):
    def __init__(self):
        super(NeuralNetwork, self).__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28*28, params['features']),
            nn.ReLU(),
            nn.Linear(params['features'], params['features']),
            nn.ReLU(),
            nn.Linear(params['features'], 10)
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits

model = NeuralNetwork().to(device)

loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=params['lr'], momentum=params['momentum'])

def train(dataloader, model, loss_fn, optimizer):
    size = len(dataloader.dataset)
    model.train()
    for batch, (X, y) in enumerate(dataloader):
        X, y = X.to(device), y.to(device)
        pred = model(X)
        loss = loss_fn(pred, y)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

def test(dataloader, model, loss_fn):
    size = len(dataloader.dataset)
    num_batches = len(dataloader)
    model.eval()
    test_loss, correct = 0, 0
    with torch.no_grad():
        for X, y in dataloader:
            X, y = X.to(device), y.to(device)
            pred = model(X)
            test_loss += loss_fn(pred, y).item()
            correct += (pred.argmax(1) == y).type(torch.float).sum().item()
    test_loss /= num_batches
    correct /= size
    return correct

epochs = 5
for t in range(epochs):
    print(f"Epoch {t+1}\n-------------------------------")
    train(train_dataloader, model, loss_fn, optimizer)
    accuracy = test(test_dataloader, model, loss_fn)
    nni.report_intermediate_result(accuracy)
nni.report_final_result(accuracy)

```

**Comment by Michaelll123** [comment]
I have got the same error, and I fixed it by degrading the nni version to 2.5

**Comment by liuzhe-lz** [comment]
Please downgrade typeguard with `pip install "typeguard<3"`.
This will be fixed in upcoming [v3.0 release](https://test.pypi.org/project/nni/3.0a0/).

**Comment by sunze992** [comment]
> Please downgrade typeguard with `pip install "typeguard<3"`. This will be fixed in upcoming [v3.0 release](https://test.pypi.org/project/nni/3.0a0

> Please downgrade typeguard with `pip install "typeguard<3"`. This will be fixed in upcoming [v3.0 release](https://test.pypi.org/project/nni/3.0a0/).

Thank you very much! It works. Looking forward to NNI v3.0 .
